elastic/kibana
# Compute dashboard panel selection list lazily

## Summary

Closes https://github.com/elastic/kibana/issues/187587

This PR changes how the dashboard panel selection items get computed, it had previously been computed eagerly, in this implementation panel selection items would only be computed when the user actually clicks the `add panel` button, with it's results cached so that subsequent interactions with the `add panel` button leverages the already computed data.

**Notable Mention:**
The options  presented as the dashboard panel list now only comprise of uiActions specifically registered with the uiAction trigger `ADD_PANEL_TRIGGER` and specific dashboard visualisation types. See https://github.com/elastic/kibana/pull/187797#discussion_r1681320456 to follow the reasoning behind this. 

That been said adding new panels to the dashboard, would be something along the following lines; 


```ts
import { ADD_PANEL_TRIGGER } from '@kbn/ui-actions-plugin/public';

 uiActions.attachAction(ADD_PANEL_TRIGGER, <registredActionId>);

// alternatively
// uiActions.addTriggerAction(ADD_PANEL_TRIGGER, ...someActionDefintion);
````

### Visuals
https://github.com/elastic/kibana/assets/7893459/7c029a64-2cd8-4e3e-af5a-44b6788faa45

### How to test 
- Navigate to a dashboard of choice
- Slow down your network speed using your browser dev tools, refresh your dashboard, and click on the “Add panel” button as soon as it is available (before the panels have a chance to load). 
- You should be presented with a loading indicator, that eventually is swapped out for the list of panels available for selection.


### Checklist

Delete any items that are not applicable to this PR.
